frontend/src/types/validate-analytics.ts

```diff
@@ -0,0 +1,70 @@
+import type { Events } from "~/types/analytics"
+
+import type { UnionToIntersection } from "type-fest"
+
+type ReservedPropNames =
+  | "timestamp"
+  | "language"
+  | "breakpoint"
+  | "ua"
+  | "os"
+  | "platform"
+  | "browser"
+  | "version"
+  | "origin"
+  | "pathname"
+  | "referrer"
+  | "width"
+  | "height"
+
+type BaseEventsType = Record<string, Record<string, string | number | boolean>>
+
+type MergePayloadTypes<E extends BaseEventsType> = UnionToIntersection<
+  E[keyof E]
+>
+type ValidatePayloads<
+  E extends BaseEventsType,
+  Payloads = MergePayloadTypes<E>
+> = Omit<Payloads, ReservedPropNames> extends Payloads
+  ? E
+  : "Payloads include reserved prop names."
```

Comment:
When I was writing the functions, the ability to override the default props was considered a feature. One such use was when I was sending a page navigation event from the router middleware where we want `pathname` to be the new page being navigated to, but `window.location.pathname` gives the current page being navigated from. So we wanted to be able to supply that and override the default.

Considering it's not used anymore we can add this restriction, I'm just adding some historical context.

Reply:
I see :thinking: From my perspective I would hope that any of the "default", "always there" props always have the same meaning for every event, rather than having to remember that their meaning can vary. If we run into absolutely needing to override the meaning of a default prop, perhaps we can deal with it then, so that we understand the use-case?